Campaign reset functionality

Hilimii opened the pull request:
I had a burst of inspiration and realised that campaign auto resets are much easier to execute then I previously thought.

The short of it:

Watch the value of `onMissionSequence `and when it changes from 0 to 1.
Also watch the value of `levelSequencePhase`. If it doesn't change at the same time as `onMissionSequence`, it means a new campaign has started, hence reset. If `levelSequencePhase `does change, it means the player has resumed a campaign, hence no reset.

Since `reset` triggers before `start` in the order of operations, it means that we can reset the timer and start it again to begin a new campaign on the same tick!

I've tested both vanilla and F59, both seem to work. If you can find any mistakes or bugs, lemme know!

Hilimii committed:
Campaign reset functionality

Adds new lines under 'reset' event to facilitate resets during whilst running the campaign.
The logic watches to when onMissionSequence changes from 0 to 1, and compares it to levelSequencePhase. If levelSequencePhase does not changes, it means a new campaign has started, thus reset, if levelSequencePhase does change, it means the player has 'resumed', thus no reset.

Hilimii committed:
Update campaign setting tooltip

Amendment to the campaign settings tooltip to indicate it has reset functionality.

Hilimii committed:
Updated version num and changelog

Changed version num to 1.0.2 and updates changelog comment to indicate the addition of reset functionality.

jwfraustro commented:
Awesome! I'll take a look this evening!

Hilimii committed:
Completed incomplete comments

Added additional comments to describe the new reset functionality.

jwfraustro commented:
Looks pretty good, but I have found one edge-case. 

If you start a new campaign and get to the first menu, with "Start Mission", return to the main menu, and hit Resume, it will reset the run. 

It's debatable whether it matters much in this case, since the time loss would only be about 40 seconds or so, but I am afraid it might pop-up elsewhere. If you can find an easy way to fix it, that would be great. Otherwise, we'll want to make sure it's pretty clearly documented that's expected.

Hilimii commented:
Ah yeah I see the potential issue.

Like you said it doesn't matter if it occurs on m1. The player is going to have to watch the pre-campaign cutscene + briefing again if they hit resume, they don't want that time counted twice in the run. The issue is if any other missions have a pre-briefing cutscene, resuming onto those missions could kill the run.

I'll do a double check over all the missions for cutscenes and test out the interaction before I merge this one. Might take me a few days and I'll report back.

Hilimii commented:
Immediate update: I suspected M17 No Respite would cause this issue because of the '2 months later' cutscene, and it did. I'll work out a way to cull it.

Hilimii committed:
Adds logic to cull undesired resets

Stops the reset event from triggering during campaign runs when resuming onto missions like m17 No respite, which have a cutscene preceding them.

Hilimii commented:
Okay I'm pretty confident that the latest push should address any future issues. It checks if the LevelID is campaign_01 or mf_01, so it doesn't falsely think a new campaign is starting when resuming onto missions like M17 No Respite, which starts with a cutscene.

Hilimii committed:
Resolving Conflicts.asl

Adds comments from the beginning of main to resolve conflicts before merging.

Hilimii committed:
Merge branch 'main' into Campaign-Reset-Test

jwfraustro commented:
Awesome, I'm glad you found that. I was sure there would be some other small "gotcha". I'll check it out again this afternoon, but if you feel good about it, I'm okay with merging it at this point.

Hilimii committed:
Version Bump

Increments version to 1.3.0 and adds patch notes.
